[RFC] Automatically retry a failed load-based auto-scale event

cegganesh84 opened the pull request:
1. Every 5 mins the scaling status of the load based auto scale enabled clusters is checked.
    2. If there is a failure in scale event then it is retried.
    3. As a side effect also retries manual scale operation failures.
    4. To prevent perpetual retries added a couple of restrictions like one retry per day for a failure which is reset upon successive auto-scaling activity.

sidseth reviewed (commented):
Does retry work for scaling flows ?
Guessing one of the intents here is to get the flow to make progress, and eventually get the cluster into a state where scaling / other operations can run again - without manual intervention?
One, in my mind fairly big risk, with an operation like this is that when a scaling flow fails / gets stuck, and if the recovery is also going to get stuck - it is going to put the cluster into a position where there is a very short window in which other repair operations like deleting nodes can be performed (this is in the absence of cancellable flows).

cegganesh84 commented:
Yeah, preventing escalations for scaling failures was the motivation. Good point on flows getting stuck for a long time. Another case where it gives a terrible experience is when there is an NPE in the scaling flow.

cegganesh84 commented:
Changed the logic to do retry only for load-based scaling. Also introduced extra checks not to retry more than once in a day, if there are no other successive scaling activities. Will close this PR, because the end-to-end testing shows that even though the request can be retried it is not fruitful in many cases where there has to be a node repair or node deletion for the upscale or downscale retry to work. The upscale cluster manager is a huge flow operation and making an idempotent retry is not possible. Will certainly be useful for the alternate mechanism of scaling where there are fewer operations and it is possible to make the operations idempotent.

cegganesh84 commented:
One other issue I found during testing. If the decommission fails then the node status becomes decommission failed, but the cluster and stack status is made available. Autoscale thinks that it should not perform further scaling operations when some nodes are in node failure state. The retry does not work because the scale operation was not marked as a failed one.
